**Rizal Adiyanto Nugroho** -
**24060122140156** -
**C1**

In [ ]:
import pandas
import numpy as np
import matplotlib.pyplot as plt
import seaborn as sns
from pandas.plotting import scatter_matrix
from sklearn import model_selection
from sklearn.neighbors import KNeighborsClassifier
from sklearn.tree import DecisionTreeClassifier
from sklearn.svm import SVC
from sklearn.naive_bayes import GaussianNB
from sklearn.metrics import classification_report
from sklearn.metrics import confusion_matrix
from sklearn.metrics import accuracy_score
from sklearn.model_selection import GridSearchCV
from sklearn.model_selection import train_test_split

In [ ]:
url = "http://archive.ics.uci.edu/ml/machine-learning-databases/iris/iris.data"
names = ['sepal-length', 'sepal-width', 'petal-length', 'petal-width', 'class']
dataset = pandas.read_csv(url, names=names)

In [ ]:
array = dataset.values
X = array[:,0:4]
Y = array[:,4]

validation_size = 0.20
seed = 7

X_train, X_validation, Y_train, Y_validation = model_selection.train_test_split(X, Y, test_size=validation_size, random_state=seed)

In [ ]:
# Test options and evaluation metric
seed = 7
scoring = 'accuracy'

In [ ]:
models = []
models.append(('KNN', KNeighborsClassifier()))
models.append(('NB', GaussianNB()))
models.append(('SVM', SVC()))
results = []
names = []
for name, model in models:
    kfold = model_selection.KFold(n_splits=10, shuffle=True ,random_state=seed)
    cv_results = model_selection.cross_val_score(model, X_train, Y_train, cv=kfold, scoring=scoring)

    results.append(cv_results)
    names.append(name)
    msg = "%s: %f (%f)" % (name, cv_results.mean(), cv_results.std())
    print(msg)

KNN: 0.983333 (0.033333)
NB: 0.966667 (0.040825)
SVM: 0.983333 (0.033333)


# **Model KNN**

In [ ]:
# Pendefinisian parameter K yang ingin diuji
param_grid = {'n_neighbors': range(1, 31)}

# Inisialisasi KNN dan GridSearch
knn = KNeighborsClassifier()
grid_search = GridSearchCV(knn, param_grid, cv=5, scoring='accuracy')
grid_search.fit(X, Y)

# Hasil optimal
print("Optimal K:", grid_search.best_params_['n_neighbors'])
print("Best Accuracy:", grid_search.best_score_)

Optimal K: 6
Best Accuracy: 0.9800000000000001


**PENCARIAN NILAI K TERBAIK**
Dimulai dengan set parameter range pencarian K pada variabel param_grid, yaitu dari 1 hingga 30. Dengan menggunakan GridSearchCV, nilai K terbaik akan dicari berdasarkan akurasi dengan 5-folds cross-validation.
Didapatkan nilai K terbaik adalah 6 dengan akurasi 98%

In [ ]:
knn = KNeighborsClassifier(n_neighbors=6)
knn.fit(X_train, Y_train)
predictions = knn.predict(X_validation)
print(accuracy_score(Y_validation, predictions))
print(confusion_matrix(Y_validation, predictions))
print(classification_report(Y_validation, predictions))

0.8666666666666667
[[ 7  0  0]
 [ 0 11  1]
 [ 0  3  8]]
                 precision    recall  f1-score   support

    Iris-setosa       1.00      1.00      1.00         7
Iris-versicolor       0.79      0.92      0.85        12
 Iris-virginica       0.89      0.73      0.80        11

       accuracy                           0.87        30
      macro avg       0.89      0.88      0.88        30
   weighted avg       0.87      0.87      0.87        30



Hasil evaluasi menunjukkan bahwa akurasi pada data validasi adalah 0.866 atau 86.6%, yang artinya 86.6% dari prediksi model sesuai dengan aktualnya.
Pada tabel confusion matrix ditunjukkan kelas Iris-setosa diprediksi sangat baik, yaitu 7 prediksi benar, dan 0 salah. Lalu Iris-versicolor dengan 11 prediksi benar, dan 1 salah ke Iris-virginica. Dan yang terakhir Iris-virginica dengan 8 prediksi benar, dan 3 salah ke Iris-versicolor.
Pada Iris-setosa menunjukkan nilai sempurna pada ketiga nilai precision, recall, dan f1-score. sedangkan pada Iris-versicolor mendapat precision sebesar 0.79, recall 0.92, dan f1-score 0.85. serta pada Iris-virginica mendapat nilai precision sebesar 0.89, recall 0.73, dan f1-score 0.80.
Rata-rata makro (macro average) dari precision, recall, dan F1-score adalah masing-masing sebesar 0.89, 0.88, dan 0.88, yang menunjukkan performa rata-rata model tanpa mempertimbangkan proporsi kelas. Sedangkan rata-rata berbobot (weighted average), yang memperhitungkan proporsi sampel dalam setiap kelas, adalah masing-masing sebesar 0.87 untuk precision, recall, dan F1-score.

# **Model Naive Bayess**

In [ ]:
nb = GaussianNB()
nb.fit(X_train, Y_train)
predictions = nb.predict(X_validation)
print(accuracy_score(Y_validation, predictions))
print(confusion_matrix(Y_validation, predictions))
print(classification_report(Y_validation, predictions))

0.8333333333333334
[[7 0 0]
 [0 9 3]
 [0 2 9]]
                 precision    recall  f1-score   support

    Iris-setosa       1.00      1.00      1.00         7
Iris-versicolor       0.82      0.75      0.78        12
 Iris-virginica       0.75      0.82      0.78        11

       accuracy                           0.83        30
      macro avg       0.86      0.86      0.86        30
   weighted avg       0.84      0.83      0.83        30



Hasil evaluasi menggunakan naive bayess menunjukkan nilai akurasi pada data adalah 0.833 atau 83.3% data terprediksi benar. Tabel confusion matrix menunjukkan prediksi pada Iris-setosa semuanya terprediksi benar, sama dengan hasil pada metode KNN. Namun pada Iris-versicolor hanya ada 9 prediksi yang benar dan 3 salah ke Iris-virginica. Serta ada 9 prediksi yang benar pada Iris-virginica dan 2 salah ke Iris-varsicolor.
Pada Iris-setosa menunjukkan nilai sempurna pada ketiga nilai precision, recall, dan f1-score. sedangkan pada Iris-versicolor mendapat precision sebesar 0.82, recall 0.75, dan f1-score 0.78. serta pada Iris-virginica mendapat nilai precision sebesar 0.75, recall 0.82, dan f1-score 0.78.
Rata-rata makro (macro average), yang merupakan rata-rata tidak berbobot dari precision, recall, dan F1-score di semua kelas, adalah masing-masing 0.86, 0.86, dan 0.83. Sedangkan rata-rata berbobot (weighted average), yang mempertimbangkan jumlah sampel dalam setiap kelas, adalah masing-masing 0.84, 0.83, dan 0.83. Dengan demikian, meskipun model bekerja dengan sangat baik pada kelas Iris-setosa, performanya sedikit menurun untuk kelas Iris-versicolor dan Iris-virginica, terutama karena adanya tumpang tindih fitur antara kedua kelas ini.